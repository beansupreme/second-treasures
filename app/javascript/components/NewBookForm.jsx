import React from "react";
import axios from "axios";
import ErrorList from "./ErrorList"
import MessageBox from "./MessageBox"

const defaultState = {
  title: '',
  author: '',
  price: 0.00,
  isbn: '',
  genre: '',
  errors: [],
  message: ''
}

class NewBookForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = defaultState

    this.handleSubmit = this.handleSubmit.bind(this);
    this.updateField = this.updateField.bind(this);
  }

  handleSubmit(event) {
    event.preventDefault();
    axios.post(
      '/api/v1/books',
      {
        book: this.buildBook()        
      }
    ).then(response => {
      this.props.onNewBookAdd(response.data);
      
      this.setState(defaultState);
      this.setState({message: `'${response.data.title}' successfully added`});
      setTimeout(() => {
        this.setState({message: ''})
      }, 5000);

    }).catch(error => {
      let response = error.response;
      if (response.status === 422) {
        this.setState({
          errors: response.data,
          message: ''
        });
      } else {
        this.setState({
          errors: ['Something went wrong... Books cannot be created at this time.'],
          message: ''
        });
      }
    })
  }

  buildBook() {
    return {
      title: this.state.title,
      author: this.state.author,
      price: this.state.price,
      isbn: this.state.isbn,
      genre: this.state.genre
    }
  }


  updateField(event) {
    let fieldName = event.target.id.split('_')[1];
    let fieldsToUpdate = {};
    fieldsToUpdate[fieldName] = event.target.value;
    this.setState(fieldsToUpdate);
  }

  render() {
    return (
      <React.Fragment>
        <h5 className="text-center">Add a new Book recommendation:</h5>
        <ErrorList id="new-book-form-errors" errors={this.state.errors} />
        <MessageBox id="new-book-form-message" message={this.state.message} />
        <form className="form-inline" id="new-book-form" onSubmit={this.handleSubmit}>
          <div className="form-group">
            <input id="book_title_field" name="title" type="text" placeholder="Title" className="form-control" 
            onChange={this.updateField} value={this.state.title}/>
          </div>
          <div className="form-group">
            <input id="book_author_field" name="author" type="text" placeholder="Author" className="form-control" 
            onChange={this.updateField} value={this.state.author}/>
          </div>
          <div className="form-group">
            <input id="book_price_field" name="price" type="number" placeholder="Price" className="form-control" 
             min="0" max="10000" step="0.01" onChange={this.updateField} value={this.state.price}/>
          </div>
          <div className="form-group">
            <input id="book_isbn_field" name="ISBN" type="text" placeholder="ISBN" className="form-control" 
            onChange={this.updateField} value={this.state.isbn}/>
          </div>
          <div className="form-group">
            <input id="book_genre_field" name="Genre" type="text" placeholder="Genre" className="form-control" 
            onChange={this.updateField} value={this.state.genre}/>
          </div>
          <button type="submit" className="btn btn-success">Save</button>
        </form>  
      </React.Fragment>
    )
  }
}

NewBookForm.defaultProps = {
  onNewBookAdd: () => {}
}

export default NewBookForm;