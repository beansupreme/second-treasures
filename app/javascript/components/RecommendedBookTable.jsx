import React from "react";
import axios from "axios";
import RecommendedBookTableRow from './RecommendedBookTableRow'

class RecommendedBookTable extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      books: []
    }
  }

  componentDidMount() {
    axios.get('/api/v1/books').then(response => {
      this.setState({books: response.data})
    });
  }

  render() {
    const recommendedBooks = this.state.books.map((book) =>
     <RecommendedBookTableRow key={book.id} book={book} />
    )
    return (
      <table id="recommended-book-table" className="table">
        <caption>Recommended Books</caption>
        <thead>
          <tr>
            <th>Title</th>
            <th>Author</th>
            <th>Price</th>
            <th>ISBN</th>
          </tr>
        </thead>
        <tbody>
          {recommendedBooks}
        </tbody>
      </table>
    )
  }
}

export default RecommendedBookTable;