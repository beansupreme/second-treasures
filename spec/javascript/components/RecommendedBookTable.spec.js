import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow} from 'enzyme';
import sinon from 'sinon';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import RecommendedBookTable from 'components/RecommendedBookTable';
import NewBookForm from 'components/NewBookForm';

describe('<RecommendedBookTable />', () => {
  let mock = new MockAdapter(axios);

  const booksResponse = [
    {
      id: 5,
      isbn: '978-1552453056',
      title: 'Fifteen Dogs',
      author: 'Andre Alexis',
      genre: 'Fiction',
      price: '17.95',
      createdAt: "2018-06-25T14:44:31.990Z",
      updatedAt: "2018-06-25T14:44:31.990Z"
    },
    {
      id: 6,
      isbn: '978-330020242',
      title: 'Vimy Ridge',
      author: 'Alexander McKee',
      genre: 'War novel',
      price: '1.50',
      createdAt: "2018-06-25T14:44:31.990Z",
      updatedAt: "2018-06-25T14:44:31.990Z"
    },
    {
      id: 7,
      isbn: '978-0060850524',
      title: 'Brave New World',
      author: 'Aldous Huxley',
      genre: 'Dystopian Fiction',
      price: '16.31',
      createdAt: "2018-06-25T14:44:31.990Z",
      updatedAt: "2018-06-25T14:44:31.990Z"
    }
  ];

  beforeEach(function() {
    mock.onGet('/api/v1/books').reply(200, 
      booksResponse
    );
  });

  it('renders as expected', () => {
    const tree = renderer.create(<RecommendedBookTable />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls to the books index endpoint on render', () => {
    sinon.spy(axios, 'get');

    const wrapper = mount(<RecommendedBookTable />);


    expect(axios.get.calledOnce).toEqual(true);

    let getData = axios.get.getCall(0).args;
    
    expect(getData[0]).toEqual('/api/v1/books');


    axios.get.restore();

  });

  xit('calls to the books delete endpoint when a row is clicked', (done) => {
    sinon.spy(axios, 'delete');

    const wrapper = mount(<RecommendedBookTable />);

    setTimeout(() => {
      let list = wrapper.find('#recommended-book-table');
      console.log(list.html())
      let firstDeleteButton = wrapper.find('#delete-book-button-5');
      console.log(firstDeleteButton)
      expect(firstDeleteButton).toHaveText('Delete');

      expect(axios.delete.calledOnce).toEqual(true);

      let getData = axios.delete.getCall(0).args;
      
      expect(getData[0]).toEqual('/api/v1/books/5');


      axios.delete.restore();
      done()
    }, 0);

  });

  it('renders the books from the get books endpoint on the page', (done) => {
    const wrapper = mount(<RecommendedBookTable />);

    setTimeout(() => {
      expect(wrapper.state('books')).toEqual(booksResponse);

      let list = wrapper.find('#recommended-book-table');
      expect(list).toIncludeText('Vimy Ridge');
      expect(list).toIncludeText('Alexander McKee');
      expect(list).toIncludeText('$17.95');
      expect(list).toIncludeText('978-1552453056');
      expect(list).toIncludeText('War novel');


      expect(list).toIncludeText('Fifteen Dogs');
      expect(list).toIncludeText('Andre Alexis');

      expect(wrapper.find('#recommended-book-table-errors').first()).toHaveHTML('<div></div>');
      
      done()
    }, 0);
  });

  it('renders errors on the page if the get books call fails', (done) => {
    mock.onGet('/api/v1/books').reply(500, 
      "Beep Boop Server error."
    );

    const wrapper = mount(<RecommendedBookTable />);

    

    setTimeout(() => {
      expect(wrapper.find('#recommended-book-table-errors').first()).toIncludeText("Something went wrong... Books are not available at this time.");
      expect(wrapper.find('#recommended-book-table-message').first()).toHaveHTML('<div></div>');
      
      done();
    }, 0);

    mock.restore();
  });

  it('renders a NewBookForm', () => {
    const wrapper = shallow(<RecommendedBookTable />);

    expect(wrapper.find(NewBookForm)).toHaveLength(1)
  });

});