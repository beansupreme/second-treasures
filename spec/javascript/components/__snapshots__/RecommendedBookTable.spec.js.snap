// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecommendedBookTable /> renders as expected 1`] = `
<table
  className="table"
  id="recommended-book-table"
>
  <caption>
    Recommended Books
  </caption>
  <thead>
    <tr>
      <th>
        Title
      </th>
      <th>
        Author
      </th>
      <th>
        Price
      </th>
      <th>
        ISBN
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
